table/block_based_table_reader.cc

```diff
@@ -2486,23 +2488,32 @@ void BlockBasedTableIterator<TBlockIter, TValue>::InitDataBlock() {
         key_includes_seq_, index_key_is_full_,
         /* get_context */ nullptr, s, prefetch_buffer_.get());
     block_iter_points_to_real_block_ = true;
+    if (read_options_.iterate_upper_bound != nullptr) {
+      data_block_within_upper_bound_ =
+          (user_comparator_.Compare(*read_options_.iterate_upper_bound,
+                                    ExtractUserKey(index_iter_->key())) > 0);
```

Comment:
I'm a little bit confused here. If the whole data block is out of bound, why do we need to creat the data block iterator at all? Can we do the check before calling NewDataBlockIterator() rather than after?

Reply:
If the check doesn't prevent the data block to be loaded. It is not very useful to me. Because it is out of bound, Next() will not be caused anymore, so what you saved is to trim this iterator from merging tree. This is trimmed for Seek() anyway, so here we are really handling a corner case, with the benefit of avoiding to re-push the iterator to the merging tree, and the iterator is already there before the Next() call.

Reply:
It is to prevent next block to be loaded. The logic is confusing because I'm reusing `data_block_within_upper_bound_` from #5111. I updated the PR to make it more clear, and delay the change here to #5111.